### ISSUE TITLE
Allow diskSetup to include partition layout

### ISSUE THREAD

**Issue by Xartos** [kubernetes-sigs/cluster-api#8524]
### What would you like to be added (User Story)?

As a cluster administrator I would like to be able to setup nodes that has a single disk with multiple partitions

### Detailed Description

Currently there's no way to have a single node VM and split its disk into multiple partitions since [this field](https://github.com/kubernetes-sigs/cluster-api/blob/527bc9f5ce9fe27256c938ba94ef0ae44a957bb4/bootstrap/kubeadm/api/v1alpha4/kubeadmconfig_types.go#L298) is only a boolean.
However, the cloud-init documentation says that this can either be a bool (to set up the disk as a single partition) or a list that describes the partition layout, [reference](https://cloudinit.readthedocs.io/en/latest/reference/examples.html#disk-setup).

So my proposal would be to allow this field to be either a boolean or a list. Alternatively add another field that describes the layout that can be merged with the existing field when generating the cloud-init user-data file.

### Label(s) to be applied

/kind feature
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

**Comment by ansh-wal** [comment]
@fabriziopandini  I can pick this up.

**Comment by sbueringer** [comment]
I'm not sure if we can make do this without a breaking change to our go API structs (which would mean we have to wait until v1beta2)

**Comment by sbdtu5498** [comment]
Oh sorry for the delayed reply @sbueringer. Had been busy with exams and stuff. Maybe by letting Layout bool there and adding a new field say LayoutSpec we can implement this feature without breaking the existing API struct. We can set default to use Layout as it is if the LayoutSpec is not provided in order to maintain backward compatibility as well.

**Comment by miltalex** [comment]
@fabriziopandini I was curious if that is something that can be done now, or there are still blockers? I would like to have a look into it when I find a bit of time.

**Comment by sbueringer** [comment]
To the best of my knowledge bool or list is not possible with crds (https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#intorstring cc @JoelSpeed). So I think we have to express this in a different way in our crd

**Comment by JoelSpeed** [comment]
Yeah that wouldn't be a structural schema, but you could represent the choices in a kube like way using a discriminated union, the yaml version would look something like:
```
diskLayout:
  type: Unpartitioned | SinglePartition | Custom
  custom:
  - sizePercentage: 0 < int < 100
    partitionType:  int (? Looking at the docs I think this is an int but not sure)
```
